Add Skeleton for ManagedMediaSource and ManagedSourceBuffer objects

jyavenard opened the pull request:
#### 3d03d36204f5a44542ca98b4df016a51b58a6305
<pre>
Add Skeleton for ManagedMediaSource and ManagedSourceBuffer objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=251974">https://bugs.webkit.org/show_bug.cgi?id=251974</a>
rdar://105206629

Reviewed by Youenn Fablet.

A “managed” MediaSource is one where more control over the MediaSource
and its associated objects has been given over to the User Agent.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp: Added.
(WebCore::ManagedMediaSource::create):
(WebCore::ManagedMediaSource::ManagedMediaSource):
(WebCore::ManagedMediaSource::buffering const):
(WebCore::ManagedMediaSource::quality const):
(WebCore::ManagedMediaSource::isTypeSupported):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h: Added.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.idl: Added.
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.cpp: Added.
(WebCore::ManagedSourceBuffer::create):
(WebCore::ManagedSourceBuffer::ManagedSourceBuffer):
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h: Added.
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.idl: Added.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::addSourceBuffer):
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::isManaged const):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
(WebCore::SourceBuffer::isManaged const):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultManagedMediaSourceEnabled):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

Canonical link: <a href="https://commits.webkit.org/260106@main">https://commits.webkit.org/260106@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/518545bc15cbec0a1cfffc6a4857b3de04a4d5bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16080 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116203 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115724 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110928 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17540 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7277 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99204 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112792 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13268 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96297 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40871 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95160 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27938 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82625 "Found 1 new API test failure: /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96509 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9192 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29289 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/95952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7154 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9784 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6291 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/95952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15353 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48851 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/104738 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6987 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11324 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25951 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp:
```diff
@@ -0,0 +1,70 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
```
Probably not correct, there should be two conditions only

jyavenard replied:
yeah, I copied the MediaSource copyright, didn't pay attention to the inner content

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
```diff
@@ -0,0 +1,60 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
```
Ditto here and maybe in other files

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.idl:
```diff
@@ -0,0 +1,60 @@
+    [CallWith=CurrentScriptExecutionContext] static boolean isTypeSupported (DOMString type);
+};
+
+ManagedMediaSource includes MediaSource;
```
Why do we need include and derivation?
Can we use composition maybe?

jyavenard replied:
Because I don't really know the difference, and I thought it was required.

I've made it pure derivation.

youennf commented on Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h:
```diff
@@ -0,0 +1,53 @@
+class ManagedSourceBuffer : public SourceBuffer {
+    WTF_MAKE_ISO_ALLOCATED(ManagedSourceBuffer);
+public:
+    static Ref<ManagedSourceBuffer> create(Ref<SourceBufferPrivate>&&, MediaSource*);
```
Can MediaSource be null? Can it be a MediaSource&?

Also shouldn't it be a ManagedMediaSource?

jyavenard replied:
I re-used the definition from SourceBuffer, I can make it a ManagedMediaSource indeed, though technically one could go with another.

jyavenard replied:
Made it take a reference.

youennf commented on Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h:
```diff
@@ -0,0 +1,53 @@
+    static Ref<ManagedSourceBuffer> create(Ref<SourceBufferPrivate>&&, MediaSource*);
+    ~ManagedSourceBuffer();
+
+    bool isManaged() const override { return true; }
```
final?

youennf commented on Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace WebCore {
+
+class ManagedSourceBuffer : public SourceBuffer {
```
final?

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
```diff
@@ -0,0 +1,60 @@
+
+namespace WebCore {
+
+class ManagedMediaSource : public MediaSource {
```
final?

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.h:
```diff
@@ -0,0 +1,60 @@
+
+    static bool isTypeSupported(ScriptExecutionContext&, const String& type);
+
+    bool isManaged() const override { return true; }
```
final?

youennf commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -697,7 +698,13 @@ ExceptionOr<Ref<SourceBuffer>> MediaSource::addSourceBuffer(const String& type)
-    auto buffer = SourceBuffer::create(sourceBufferPrivate.releaseReturnValue(), this);
+    Ref<SourceBuffer> buffer =
+#if ENABLE(MANAGED_MEDIA_SOURCE)
+        isManaged() ? ManagedSourceBuffer::create(sourceBufferPrivate.releaseReturnValue(), this).get() : SourceBuffer::create(sourceBufferPrivate.releaseReturnValue(), this).get();
```
If we want a ManagedMediaSource, I would tend to add the downcast infra

jyavenard replied:
I was thinking about that too, good point

youennf commented on Source/WebCore/Modules/mediasource/ManagedMediaSource.idl:
```diff
@@ -0,0 +1,60 @@
+    attribute EventHandler onstartstreaming;
+    attribute EventHandler onendstreaming;
+
+    [CallWith=CurrentScriptExecutionContext] static boolean isTypeSupported (DOMString type);
```
```suggestion
    [CallWith=CurrentScriptExecutionContext] static boolean isTypeSupported(DOMString type);
```

jyavenard replied:
This is the syntax as used in MediaSource.idl and the spec one (always a space)

jyavenard committed:
Add Skeleton for ManagedMediaSource and ManagedSourceBuffer objects
https://bugs.webkit.org/show_bug.cgi?id=251974
rdar://105206629

Reviewed by Youenn Fablet.

A “managed” MediaSource is one where more control over the MediaSource
and its associated objects has been given over to the User Agent.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediasource/ManagedMediaSource.cpp: Added.
(WebCore::ManagedMediaSource::create):
(WebCore::ManagedMediaSource::ManagedMediaSource):
(WebCore::ManagedMediaSource::buffering const):
(WebCore::ManagedMediaSource::quality const):
(WebCore::ManagedMediaSource::isTypeSupported):
* Source/WebCore/Modules/mediasource/ManagedMediaSource.h: Added.
* Source/WebCore/Modules/mediasource/ManagedMediaSource.idl: Added.
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.cpp: Added.
(WebCore::ManagedSourceBuffer::create):
(WebCore::ManagedSourceBuffer::ManagedSourceBuffer):
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.h: Added.
* Source/WebCore/Modules/mediasource/ManagedSourceBuffer.idl: Added.
* Source/WebCore/Modules/mediasource/MediaSource.cpp:
(WebCore::MediaSource::addSourceBuffer):
* Source/WebCore/Modules/mediasource/MediaSource.h:
(WebCore::MediaSource::isManaged const):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
(WebCore::SourceBuffer::isManaged const):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultManagedMediaSourceEnabled):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:

Canonical link: https://commits.webkit.org/260106@main
